File: 3163-string-compression-iii/3163-string-compression-iii.kt
class Solution {
    fun compressedString(word: String): String = buildString {
        var i = 0
        while (i < word.length) {
            var j = i
            while (j < minOf(i + 9, word.length) && word[i] == word[j]) j++
            append("${j - i}${word[i]}")
            i = j
        }
    }
}

File: 3163-string-compression-iii/3163-string-compression-iii.rs
impl Solution {
    pub fn compressed_string(word: String) -> String {
        word.into_bytes()
            .chunk_by(|a, b| a==b)
            .flat_map(|chunk| {
                chunk
                    .chunks(9)
                    .flat_map(|c| [Into::<char>::into(c.len() as u8 + b'0'), chunk[0].into()])
            })
            .collect()
    }
}
